Kuzmin Turek implementation for advection with constant velocity

WilkAndy opened the pull request:
Please comment and advise on how to do this properly, @joshuahansel , @permcody , @andrsd 

a

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

WilkAndy committed:
Kuzmin Turek works !!   This is only for 1D with various simplifying assumptions.  Refs #12370

joshuahansel commented:
I still don't think FCT is even possible until someone comes up with a way to do multiple solves in a time step, since the antidiffusive fluxes generally depend on the high-order solution. An exception would be if both of the following are true:
1. The high-order scheme uses a lumped mass matrix.
2. Fully explicit time integration is used.

In this case, the anti-diffusive fluxes would only depend on the old solution. I don't know how great the overall scheme would be if 1. is satisfied.

WilkAndy commented:
OK, thanks for your comments on FCT, @joshuahansel .  Today i want to do a worked example (by hand) that will hopefully convince me of whether FCT is going to be possible and worthwhile.

WilkAndy commented:
I believe that Kuzmin-Turek works.  I'm really excited.  There's a lot of work to do, probably both in the framework and PorousFlow, but i'm attaching a figure illustrating the exciting results.  I hope that @hsheldon is as please as i am, and if we get this working for the general case, i think it'll be an extremely well-received addition.



@joshuahansel , i don't think two separate solves are needed.  Instead steps (1) to (3) in KT Fig 2 just modify the advection Kernel, by adding diffusion (D) and anti-diffusion (F).  This is what is implemented in the attached PR.  Because of the stuff below, i have only been able to implement KT in 1D for constant velocity where i can compute their "K" matrix by hand.

@andrsd , @permcody , @joshuahansel , here's what i think we need to make this general:
 - Ability to compute KT's K matrix, given in their Eqn(20).  This looks very much like our usual residuals, save for lack of multiplication by _u[_qp] in the integral, so it can almost be computed by a Kernel.  However, the key is that k_ij must be computed for all linked pairs of nodes, and be stored somewhere so that the remainder of the algorithm can use it.  We can certainly compute it by looping over elements, and just increment the appropriate k_ij entries.  Almost certainly, the velocity, v, will be a MaterialProperty in PorousFlow, so it'll have to go into the integral.  We will also need derivatives of k_ij wrt to all the variables.
 - Ability to get a list of all the nodes connected to a certain node.  These appear in the sums over "j" in KT.
 - Ability for the Kernel to access the k_ij, given global node numbers of the element (or similar), as well as nodal values of u.  Probably u will be a "nodal" Material, like many other things in PorousFlow.

I'd like your help with this, please.
![fltvd_comparison](https://user-images.githubusercontent.com/6676821/47426687-3f9a1780-d7d1-11e8-9fc1-a170c96d422f.png)

WilkAndy commented:
I forgot to mention, the figure is generated in the same way as the final figure in the recent numerical_diffusion doco (http://mooseframework.inl.gov/docs/PRs/12348/site/modules/porous_flow/numerical_diffusion.html)

joshuahansel commented:
Those results look great. It appears that the algorithm that you and KT are using is the different from the FCT algorithm I know. I've read the paper a little, but I think it would take a longer time investment for me to truly understand it all. Specifically, the issue of confusion is the computation of the anti-diffusive fluxes. As I had known it, the anti-diffusive fluxes, when applied without limitation, were supposed to reproduce the high-order solution - this property yielded the scheme that I had been using. It appears in KT's paper that the anti-diffusive fluxes were not designed around this property, and thus they did not need to depend on the high-order solution (I guess they just depend upon the implicit solution?).

I think your questions at this point are MOOSE questions, so I'll leave that to the MOOSE experts, as I need to spend my time working on RELAP-7. I'd be very much interested to see what can be done further with this work though! Clearly, I need to read some more on FCT sometime...

andrsd commented:
> Ability to compute KT's K matrix, given in their Eqn(20).  This looks very much like our usual residuals, save for lack of multiplication by _u[_qp] in the integral, so it can almost be computed by a Kernel.  However, the key is that k_ij must be computed for all linked pairs of nodes, and be stored somewhere so that the remainder of the algorithm can use it.  We can certainly compute it by looping over elements, and just increment the appropriate k_ij entries.

Sounds like a job for `UserObject`. You should be able to get all the linked nodes via libMesh.  For serial runs this will be easy. When you get to implement the parallel version, you will probably need to use the relationship manager @permcody mentioned earlier, so that you can get the right level of ghosting.

> Ability to get a list of all the nodes connected to a certain node. These appear in the sums over "j" in KT.

This sounds related to (1) - it should not be a problem

> Ability for the Kernel to access the k_ij, given global node numbers of the element (or similar), as well as nodal values of u.

You can design your UserObject (as I mentioned above) such that you can request the data the way you need them. 

Getting nodal values of variables has been in the framework for a long time - this should not a be a problem - look up `coupledNodalValue` in `Coupleable` interface.

> Probably u will be a "nodal" Material, like many other things in PorousFlow.

I am not sure I understand this. We solve for `u`, right? So why nodal material?

And just out of curiosity, you want KT because you want to stick with continuous Galerkin?

WilkAndy commented:
> Sounds like a job for `UserObject`. You should be able to get all the linked nodes via libMesh. For serial runs this will be easy. When you get to implement the parallel version, you will probably need to use the relationship manager @permcody mentioned earlier, so that you can get the right level of ghosting.

Yes, i used `UserObject` in the current implementation, but i didn't know i could get the linked nodes via libMesh (i just assumed the dofs at the nodes were ordered from left-to-right along my 1D line).  Could you point me to some methods, please?  Libmesh is quite opaque to me.

> 
> > Ability to get a list of all the nodes connected to a certain node. These appear in the sums over "j" in KT.
> 
> This sounds related to (1) - it should not be a problem
> 
> > Ability for the Kernel to access the k_ij, given global node numbers of the element (or similar), as well as nodal values of u.
> 
> You can design your UserObject (as I mentioned above) such that you can request the data the way you need them.
> 
> Getting nodal values of variables has been in the framework for a long time - this should not a be a problem - look up `coupledNodalValue` in `Coupleable` interface.

I had to screw around a little in the current implementation, probably because of the lack of knowing what the neighbouring nodes are.  My `Kernel` will ask: "what is the flux correction for node such-and-such" and my `UserObject` will then need to compile a list of neighbouring nodes, pull out K_ij and the value of "u" at those nodes, compute and return the final result.

I couldn't do that with coupledNodalValue because of not knowing what the neighbours were.

> 
> > Probably u will be a "nodal" Material, like many other things in PorousFlow.
> 
> I am not sure I understand this. We solve for `u`, right? So why nodal material?
> 

This is just because of the architecture of PorousFlow: most things are nodal Materials.  But if you like you can think of "u" as a nonlinear function of the `Variables` of the problem: it is exactly the same as a nodal Material, but in a different guise.

> And just out of curiosity, you want KT because you want to stick with continuous Galerkin?

Yes, primarily.  In many parts of PorousFlow we assume that the Variables are linear Lagrange.  (Actually, the KT idea may allow us to get around a lot of this.)  RDG clashes with a lot of PorousFlow and we'll need a lot of subtle changes to integrate those ideas into PorousFlow.  I suspect that KT and RDG(P0P1) yield very similar results.

WilkAndy commented:
@joshuahansel - yes, the antidiffusive fluxes are cleverly based on how much heat is moving to this node and how much is moving away.  So you don't need to solve anything before you compute the antidiffusion - you just base your computation on the current solution!   In 1D it's doing *exactly* the same as RDG(P0P1), but swap nodes with elements.  In higher dimensions it's probably the same too, but i haven't checked by hand.

WilkAndy commented:
Don't worry, @nadrsd , google helped me.  I found this thread, which i presume is still correct:

https://sourceforge.net/p/libmesh/mailman/message/34587502/

a

WilkAndy committed:
FluxLimiter uses find_nodal_neighbors

WilkAndy commented:
Hi @andrsd .  So i played around with this for a few hours today.  i made a little progress utilizing the neighbouring nodes, thanks.

In my `GeneralUserObject` i have a loop over elements:
```
 for (const auto & elem : _mesh.getMesh().active_element_ptr_range())
```
Within this loop, i'd like to do this sort of thing:
```
  for (unsigned i = 0; i < num_nodes; ++i)                                                                                                                                                                          
    for (unsigned j = 0; j < num_nodes; ++j)                                                                 
      for (unsigned qp = 0; qp < _qrule->n_points(); ++qp)                                                   
        kk[i][j] += _JxW[qp] * _coord[qp] * (_grad_test[i][qp] * _velocity) * _test[j][qp];  
```
but `GeneralUserObject` doesn't know `_qrule`, `_JxW`, `_coord`, `_grad_test` or `_test`.  What's a good way for `GeneralUserObject` to access these things ?

andrsd commented:
> `GeneralUserObject` doesn't know `_qrule`, `_JxW`, `_coord`, `_grad_test` or `_test`

Yeah, because it is not associated with any mesh entity. I made a special mesh sweeping class that derives from `GeneralUserObject` to accommodate the rDG needs - it is in the rDG module and is called `ElementLoopUserObject`. That will give you `_JxW`, `_coord` and `_q_rule`, but not `_test` or `_grad_test`, because we did not need them. But, it provides bunch of virtuals that you can use to do what you need.  The one you want to play with are `computeElement`, `computeBoundary`, `computeInternalSide`. But for your case, you probably want to put some code inside `computeElement`.  You will need to this as first:
```
_fe_problem.prepare(_current_elem, _tid);
_fe_problem.reinitElem(_current_elem, _tid);
```
that will interpolate the variable values and reinit test functions, gradients, etc.

If your user object acts on a variable, you will probably want to use this:
```
    MooseVariableInterface<Real>(this,
                                 false,
                                 "variable",
                                 Moose::VarKindType::VAR_NONLINEAR,
                                 Moose::VarFieldType::VAR_FIELD_STANDARD),
    _var(*mooseVariable()),
    _test(_var.phi()),
    _grad_test(_var.gradPhi()),
    _phi(_assembly.phi(_var)),
    _grad_phi(_assembly.gradPhi(_var)),
```
That is from `Kernel.C` (go look there for more details). That will give you the usual stuff you are used to.  If you do not have a variable, there is another way to get `_test` and `_grad_test`:
```
    _phi(_assembly.fePhi<Real>(_fe_type)),
    _grad_phi(_assembly.feGradPhi<Real>(_fe_type)),
```
This is tailored to your specific problem - `Real` means scalar-valued FE-type and for `_fe_type`, you might have something like `(FIRST, LAGRANGE)`.

If you go with the `MooseVariable` approach, you will get one sweep through mesh for each instance of that user object and that's probably not what you want when you move to the implementation for porous flow. You will want one sweep and take care of all the caching at once. So you need to decide how to get the `_fe_type` in the user object.  One way might be, that you provide a list of variables you will be operating on, you check that all have the same `FEType` and that would be the one you use. I do not think you want to combine `FEType`s...

andrsd commented:
> can you give me some idea of how to get all nodes attache to a certain node?

There is also `MooseMesh::nodeToElemMap()` that will give you all elements connected to a node. Then you can loop over nodes of each element skipping the one you passed into the `nodeToElemMap` call...

WilkAndy commented:
Thanks for your answers about `_JxW`, etc, @andrsd .   I think i've got the whole day to work on this!

WilkAndy committed:
Renamed FluxLimiter to AdvectiveFluxCalculator
AdvectiveFluxCalculator now depends on ElementLoopUserObject in the RDG module in order to do the loop over all elements to compute Kij
In-code doco tidied

WilkAndy commented:
Hi @andrsd , @permcody , @joshuahansel .   I think this PR would benefit from a brief review.  You're probably short on time, so please concentrate your efforts on the `AdvectiveFluxCalculator` `UserObject`, as the remainder of the stuff is either doco or pretty straightforward Kernel stuff.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/250118/) on 2aa5037 wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/12371/site)

This comment will be updated on new commits.

WilkAndy commented:
I've realised that in multiple dimensions, the Kernel must divide it's residual contributions by the valence of each edge.  Viz, if an edge between two nodes is shared by N elements, then the Kernel's contribution must be divided by N.  The current Kernel implementation loops over all elements, and adds the residual contributions N times.  All i need do is to divide K_ij by the valence of the i-j edge.

A faster and cleaner way would be for the Kernel to loop over nodes, computing the residual contributions directly from all connected nodes (using K_ij).  Probably not very MOOSian, however.

andrsd commented:
BTW @WilkAndy : why did you use `GeneralUserObject`? `ElementUserObject` would have all the `_JxW`, `_coord`, variable values, etc. for you.

andrsd commented on modules/porous_flow/src/userobjects/AdvectiveFluxCalculator.C:
```diff
@@ -0,0 +1,242 @@
+    // only need to do that at the beginning and when the mesh has changed because of adaptivity
+    // (and anything else i haven't thought about?)
+    // QUERY: does this properly account for multiple processors?
+    // QUERY: or do we want to use _mesh.nodeToElemMap() ?   (if so, then
```
Yes, this is already in MooseMesh, so use that instead of solving that problem again

WilkAndy replied:
I can use `nodeToElemMap` (which returns a `std::map< dof_id_type, std::vector< dof_id_type > >` )but then i can't see how i can use `MeshTools::find_nodal_neighbors` since the latter requires a `std::vector<std::vector<const Elem *>>`.  I don't really care about finding neighbors next to nodes: i just want to find the nodal neighbors, but the elements seem to be a necessary intermediate step

andrsd replied:
I misunderstood line 74 and 75 below. I thought you were trying to build what we already have in MooseMesh. On second look, it is not the case, so your code is fine.

But, you will want to change the local variable name from `_nodes_to_elem_map` to `nodes_to_elem_map` :-)

andrsd commented on modules/porous_flow/src/userobjects/AdvectiveFluxCalculator.C:
```diff
@@ -0,0 +1,242 @@
+      const Node & node_j = _current_elem->node_ref(j);
+      for (unsigned qp = 0; qp < _qrule->n_points(); ++qp)
+        // KT Eqn (20)
+        _kij[node_i][node_j] +=
```
Why not `node_i->id()`? You are using the whole `Node` class as an index + doing a memory copy here.

WilkAndy replied:
Thanks for all your comments, @andrsd .    For this one (use `node_i->id()`), are you suggesting my _kij map should be
```
std::map<dof_id_type, std::map<dof_id_type, Real>>
```
instead of `std::map<Node, std::map<Node, Real>>` ?

dschwen replied:
I wonder if `std::map<std::pair<dof_id_type, dof_id_type>, Real>` would be even better here.

WilkAndy replied:
Decided against the std::pair idea.  I now use _kij to figure out the connected nodes, so the `std::map<dof_id_type, std::map<dof_id_type, Real>>` is better.  But i'm going to change _valence to use the std::pair.

andrsd commented on modules/porous_flow/include/userobjects/AdvectiveFluxCalculator.h:
```diff
@@ -0,0 +1,116 @@
+template <>
+InputParameters validParams<AdvectiveFluxCalculator>();
+
+class AdvectiveFluxCalculator : public ElementLoopUserObject
```
Use `ElementUserObject` if possible. Not sure you have any limitations that would prevent it. From looking at your code briefly, it should be doable.

andrsd replied:
You should also be able to compute the edge valency in this user object and then you would pull it out into the kernel via your defined API.

WilkAndy replied:
I'm not sure why you're proposing `ElementUserObject`, @andrsd .  I want to do a loop over elements, computing my `_kij` and your `ElementLoopUserObject`, with its `pre()` and `computeElement()` methods seem to be exactly what i want.

andrsd replied:
`ElementUserObject::execute()` is called on every element, that is equivalent to `ElementLoopUserObject::computeElement()` except all reinits are done for you so you do not have to deal with them. The `pre()` method would be equivalent to `ElementUserObject::initialize()`.

The difference between the 2 is that all `ElementUserObjects` are executed in one mesh sweep. While `ElementLoopUserObjects` will do 1 mesh sweep per instance.

andrsd commented on modules/porous_flow/include/userobjects/AdvectiveFluxCalculator.h:
```diff
@@ -0,0 +1,116 @@
+  RealVectorValue _velocity;
+
+  /// the nodal values of u
+  MooseVariable * _u_nodal;
```
If use decide to try the `ElementUserObject`, you would have `coupledNodalValue` API available which would be more MOOSEian.

andrsd commented on modules/porous_flow/src/kernels/FluxLimitedTVDAdvection.C:
```diff
@@ -0,0 +1,148 @@
+    for (unsigned j = 0; j < num_nodes; ++j)
+      // negative sign because residual = -Lu (KT equation (19))
+      _flux_out(i) -= ll[i][j] * _u_nodal[j] + fa[i][j];
+  return;
```
Useless line 🤪

andrsd commented on modules/porous_flow/src/userobjects/AdvectiveFluxCalculator.C:
```diff
@@ -0,0 +1,242 @@
+      nodal_flux.emplace(*n, 0.0);
+    nodal_flux.emplace(*node_i, 0.0); // include node_i - node_i connection
+    _kij.emplace(*node_i, nodal_flux);
+  }
```
Why would not you just traverse through `_kij` and set its elements to zero? Are you trying to preallocate the data structure? Because you could build up the data structure on the first pass and the subsequent ones would just zero it out. Just an idea.

andrsd replied:
If you want to stick with this though, look at `MooseMesh::localNodesBegin` and `MooseMesh::localNodesEnd`.

permcody commented:
Look on libMesh’s Mesh or MeshBase object. There are methods that will give
you a Node or Elem object if you have the ID. No searching required!
On Tue, Oct 30, 2018 at 7:29 PM Andy Wilkins <notifications@github.com>
wrote:

> *@WilkAndy* commented on this pull request.
> ------------------------------
>
> In modules/porous_flow/src/userobjects/AdvectiveFluxCalculator.C
> <https://github.com/idaholab/moose/pull/12371#discussion_r229540027>:
>
> > +  // TODO: test this
> +  if (!_execute_enum.contains(EXEC_TIMESTEP_BEGIN) || !_execute_enum.contains(EXEC_NONLINEAR))
> +    mooseError("The AdvectiveFluxCalculator UserObject " + name() +
> +               " execute_on parameter must include, at least, 'timestep_begin nonlinear'");
> +}
> +
> +void
> +AdvectiveFluxCalculator::timestepSetup()
> +{
> +  if (_nodes_to_elem_map.size() == 0)
> +  {
> +    // QUERY: We don't want to recompute the neighboring nodes every nonlinear iteration.  We
> +    // only need to do that at the beginning and when the mesh has changed because of adaptivity
> +    // (and anything else i haven't thought about?)
> +    // QUERY: does this properly account for multiple processors?
> +    // QUERY: or do we want to use _mesh.nodeToElemMap() ?   (if so, then
>
> I can use nodeToElemMap (which returns a std::map< dof_id_type,
> std::vector< dof_id_type > > )but then i can't see how i can use
> MeshTools::find_nodal_neighbors since the latter requires a std::vector<std::vector<const
> Elem *>>. I don't really care about finding neighbors next to nodes: i
> just want to find the nodal neighbors, but the elements seem to be a
> necessary intermediate step
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/12371#discussion_r229540027>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AC5XII_hnxxWJ5t3t75Gg6jo2-HrHvl7ks5uqPzdgaJpZM4X03RE>
> .
>

WilkAndy commented:
Thanks @permcody .   I see  `virtual const Node& libMesh::MeshBase::node_ref  (  const dof_id_type  i  ) ` which should do the trick.
In writing this code, i've been a bit worried about all this node and element numbering.  Eg, how do i tell that `i` in the above method is actually equivalent to `node.id()`.  Maybe it's equivalent to `node.unique_id()` or one of the other types of "id".

WilkAndy committed:
_kij now uses dof_id_type instead of Node.

Various rearrangements to keep the use of _nodes_to_elem_map minimised: once we've built _kij i can use that to determine the nodal connections

WilkAndy committed:
AdvectiveFluxCalculator now calculates valence, which is used by FluxLimitedTVDAdvection
FluxLimitedTVDAdvection now computes some Jacobian information

WilkAndy committed:
AdvectiveFluxCalculator _valence now uses std::pair, and its computation and initializing of _kij are now combined

Some tests added

WilkAndy committed:
Lots of Jacobian stuff added that makes code awful

I believe, but have NOT CHECKED, that the Jacobian is now correct for all flux_limiter_types, once i can add (in FluxLimitedTVDAdvection::computeJacobian) all the contributions that have been calculated: currently only the contributions local to the current element are added

Refs #12370

WilkAndy commented:
I'll be away for much of November, so i've asked @hsheldon whether she might like to add some tests and/or doco here.  In no particular order, this is what i would like to do:

- [ ] Test that the scheme works on different meshes.  When i say "the scheme works", i mean, observing that the code just doesn't crash, and attempting to see that the advection speed is as desired.  The meshes include
  - 2D and 3D meshes as well as the 1D i've explored already
  - meshes with different blocks
  - meshes with different blocks where the "u" is only defined on a subset of those blocks
  - meshes with not just quad elements, but some triangles too

 - [ ] Test that the scheme works with different number of processors.  I just want to know whether the results are the same as the single-proc case, because all this screwing around with MOOSE at a lower level than usual might have messed something up.

 - [ ]  Test that the scheme works with mesh adaptivity.

 - [ ] Test that the scheme works with multiple threads.

 - [ ]  Add the full Jacobian stuff that is computed in `_dflux_out_dvar` in `FluxLimitedTVDAdvection::computeJacobian()`.  At present only the contributions local to the current element are added.  Verify (hope) that these contributions are correct in some tests.

 - [ ]  Add another 2 cases to the numerical_diffusion documentation (http://mooseframework.org/moose/modules/porous_flow/numerical_diffusion.html) one with flux_limiter_type=None (which should be identical to the full-upwind case) and one with flux_limiter_type=superbee (or whatever) that should be identical to the RDG(P0P1) case.

 - [ ]  Complete the kt_worked.md documentation

 - [ ]  Document (in a separate page) the effect of different flux_limiter_types, in a similar way to the example on the RDG page: http://mooseframework.org/moose/modules/rdg/index.html

WilkAndy committed:
documentation of new Kernel and UserObject and kt_worked complete

Refs #12370

WilkAndy commented:
@hsheldon , the test failures are wierdo ones that i think aren't related to this PR

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/251313/) on 2b972eb : invalidated by @milljm

password

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/251297/) on 12d3940 : invalidated by @milljm

password updated

moosebuild commented:
Job [Modules debug PETSc alt](https://civet.inl.gov/job/251327/) on 2b972eb : invalidated by @WilkAndy

Apparently unrelated failures such as: tensor_mechanics/test:line_material_rank_two_sampler.rank_two_sampler: Assertion failed in file ../src/mpid/ch3/channels/nemesis/netmod/tcp/socksm.c at line 596: hdr.pkt_type == MPIDI_NEM_TCP_SOCKSM_PKT_ID_INFO || hdr.pkt_type == MPIDI_NEM_TCP_SOCKSM_PKT_TMPVC_INFO tensor_mechanics/test:line_material_rank_two_sampler.rank_two_sampler: internal ABORT - process 0 tensor_mechanics/test:line_material_rank_two_sampler.rank_two_sampler: [unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1

WilkAndy committed:
AdvectiveFluxCalculator now calculates and caches all residual and Jacobian info that is used by the FluxLimitedTVDAdvection Kernel

In the case of flux_limiter_type = None this is a complete and nicely inexpensive way of computing the required information

In the case of other flux_limiter_types, I believe that the Jacobian info is correctly calculated.  There are two problems:

 (1) (minor) when the "u" at all the nodes are all equal then the PETSc snes test gives different results than the analytically coded expressions.  I need to check this out thoroughly but it's probably simply a problem of going down different "if (u_i < u_j)" branches, which will be unimportant in real simulations.

 (2) (major) I think all the Jacobian entries are computed correctly, including those encoding the change of the residual at a node wrt "u" at two nodes away.  However, these contributions are not added into PETSC's Jacobian in computeJacobian, because they come from a nonstandard sparsity.  So, i need to work out how to add them properly.  By explicitly outputting these contributions (see commented-out section at the end of the post() method) and comparing with PETSc's finite differences for the jacobian_02.i input file, it may be checked that the contributions are actually correct (i suppose i could have just been lucky in the few cases i've checked).

Refs #12370

WilkAndy committed:
all Jacobian contributions now included

WilkAndy commented:
Hi @andrsd , i'd like ideas on how to make this latest commit better.  As you can see, i made `addJacobianBlock` of `Assembly` public, and i'm betting you guys don't like that.  I use it in `FluxLimitedTVDAdvection` to add all the Jacobian entries (next-nearest neighbours as well as nearest-neighbours)

WilkAndy commented:
Hi @andrsd .  I'm back from my trip and would like to start working on this again.  Could you please comment on:

Hi @andrsd , i'd like ideas on how to make this latest commit better. As you can see, i made addJacobianBlock of Assembly public, and i'm betting you guys don't like that. I use it in FluxLimitedTVDAdvection to add all the Jacobian entries (next-nearest neighbours as well as nearest-neighbours)

andrsd commented:
I will be able to get to this next week - too many other things that have higher priority.

WilkAndy commented:
Cool, thanks @andrsd .   I will work on other things in the meantime.  But i don't want to leave this too long as i have to complete the whole thing - PorousFlow version - by Christmas

WilkAndy commented:
Just reminding @andrsd to comment on the commit above

WilkAndy committed:
FluxLimited TVD cases added to numerical_diffusion documentation.

Tests (largely due to Heather Sheldon, thank you Heather) added that demonstrate correct behaviour in 1D, 2D and 3D, and with more than 1 block, and with non-quad elements, and with mesh adaptivity.

Test added that demonstrates correct Jacobian for superbee

kt_worked.md minor changes
numerical_diffusion.md now includes presentation of KT results (thank you Heather Sheldon)

Refs #12370

andrsd commented on modules/porous_flow/src/kernels/FluxLimitedTVDAdvection.C:
```diff
@@ -64,29 +64,45 @@ FluxLimitedTVDAdvection::computeJacobian()
+      deriv_matrix(0, j) = node_j_deriv.second / valence;
+      j++;
     }
+    // Ad the result to the system's Jacobian matrix
```
Ad -> Add

WilkAndy commented on modules/porous_flow/src/kernels/FluxLimitedTVDAdvection.C:
```diff
@@ -64,29 +64,45 @@ FluxLimitedTVDAdvection::computeJacobian()
+    // Ad the result to the system's Jacobian matrix
+    // TODO: get tag properly in _sys.getMatrix(0) properly
+    _assembly.addJacobianBlock(
+        _sys.getMatrix(0), deriv_matrix, idof_indices, jdof_indices, _var.scalingFactor());
```
That works, thanks @andrsd

WilkAndy committed:
FluxLimitedTVDAdvection uses cacheJacobianBlock so that Assembly::addJacobianBlock can be made protected as it previously was

AdvectiveFluxCalculator now inherits from ElementUserObject instead of ElementLoopUserObject
AdvectiveFluxCalculator attempts to do threadJoin
AdvectiveFluxCalculator now has vanLeer as its default because that is smooth so probably better convergence
AdvectiveFluxCalculator attempts to use RelationshipManager

Some more tests added

Refs #12370

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/260937/) on ff8de11 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/12371/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/12371/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 310861fa4c4e956a056662619111e8e549c08ea2`

WilkAndy committed:
large_media updated with flux_limiter_choice.gif.  Refs #12370

kt_worked documentation now contains a description of the different flux limiters, and an animation of their differences

Refs #12370

AdvectiveFluxCalculator attempting ghosting and mpi comm.  Refs #12370

start to create some virtual methods that will be overridden for porousflow.  Refs #12370

make AdvectiveFluxCalculatorBase.  Refs #12370

added multi-processor notes to AdvectiveFluxCalculator

Note that multiproc stuff does NOT work yet because of bug(?) in RelationshipManagers stuff

Refs #12370

large_media updated

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit c1fca1e68947482c7ea4ed4525b6342a24d07647
+Subproject commit 14afa5fae7b36c034daa29878b1e35931b8487c1
```
**Caution!** This contains a submodule update

WilkAndy committed:
large_media conflict

WilkAndy committed:
formatting

WilkAndy commented:
Hi @andrsd .   I think it is appropriate to merge this if it passes and if you see no problems.  I have not yet got the multi-processor case working (#12557 #12558), but the PR is growing.  Next i need to do the PorousFlow bit, so now is a natural break point.

moosebuild commented:
Job [Pthreads Test](https://civet.inl.gov/job/267213/) on 628197d : invalidated by @WilkAndy

all sorts of weirdo errors

WilkAndy commented:
Thanks for all those comments on the Kuzmin-Turek stuff, @andrsd .  It'll be a few days before i can address them, so you're off the hook for a while :-)

a

WilkAndy commented on modules/porous_flow/src/userobjects/AdvectiveFluxCalculatorBase.C:
```diff
@@ -0,0 +1,700 @@
+AdvectiveFluxCalculatorBase::getdFluxOutdu(dof_id_type node_i) const
+{
+  const auto & row_find = _dflux_out_du.find(node_i);
+  mooseAssert(row_find != _dflux_out_du.end(),
```
I'm not sure about this, @andrsd.  I don't think the user can ever trigger this Assert by making a weirdo input file or mesh.  I think the only way of triggering the Assert is by poor C++ coding, ie, mainly the Assert is for me alone.  So i think an Assert is better than something that's going to be in the optimized code.   Do you agree?

andrsd replied:
I was thinking that for the case when people (developers) will be using your API, it would be nice to guard it, because they might get into weird problems. So it is better to error out where the actual error is. We had a similar problem in RELAP and the code was bombing out in a different spot and you where not able to tell. So I like to error out as soon as possible.

However, if you do not think it is a problem, keep it this way.

WilkAndy replied:
I think your experience trumps my belief

WilkAndy commented on modules/porous_flow/include/userobjects/AdvectiveFluxCalculatorBase.h:
```diff
@@ -0,0 +1,201 @@
+   * @return result whose keys are node ids connected to node_i (according to _kij) and whose values
+   * are zero
+   */
+  std::map<dof_id_type, Real> zeroedConnection(dof_id_type node_i) const;
```
I think i do want a copy of the map in this case.

andrsd replied:
Then, you want to document it in the comment, because that will be the first thing people will ask ("why are you returning a copy").

WilkAndy replied:
I've changed this to
```
void zeroedConnection(std::map<dof_id_type, Real> & the_map, dof_id_type node_i);
```
The method clears and appropriately initializes `the_map`

WilkAndy committed:
changes requesed by andrsd

Fixes #12370

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/268836/) on 9e4cb57 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/12371/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/12371/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 310861fa4c4e956a056662619111e8e549c08ea2`

WilkAndy commented:
All set to merge?  I think i've addressed all of @andrsd 's comments

WilkAndy committed:
andrsd changes: make zeroedConnection not produce a copy of a std::map.  Use mooseError instead of mooseAssert.

Fixes #12370

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/271063/) on d6709a7 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/12371/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/12371/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 310861fa4c4e956a056662619111e8e549c08ea2`

WilkAndy commented:
Thanks for all the careful inspections and advice on this PR, @andrsd
